### ISSUE TITLE
Mapsui 4.0 roadmap

### ISSUE THREAD

**Issue by pauldendulk** [Mapsui/Mapsui#906]
Considering the roadmap voor Mapsui 4.0.

Some options:
- Use NTS
- Migrate to MIT (Use NTS and other changes)
- Improve performance. Measure performance. Transform to SkiaSharp objects when fetching data.
- VectorTile support (could be through library).
- World Wrap https://github.com/Mapsui/Mapsui/issues/518

It is possible to vote with a thumbs up on feature-request issues:
https://github.com/Mapsui/Mapsui/issues?q=is%3Aopen+label%3Aenhancement+sort%3Areactions-%2B1-desc

I move some of the feature request issues to a backlog repo https://github.com/Mapsui/backlog/issues (not sure if that was a good idea).

**Comment by pauldendulk** [comment]
Replacing the Mapsui geometries is a pre-requisite for going to MIT. Options:
- Use NTS
- Create something simple ourselves.

NTS might have some additional benefits
- Projections:  https://github.com/NetTopologySuite/ProjNet4GeoAPI
- Shapefile support: https://github.com/NetTopologySuite/NetTopologySuite.IO.ShapeFile
- VectorTile writer: https://github.com/NetTopologySuite/NetTopologySuite/issues/379

**Comment by charlenni** [comment]
I used NTS for the vector tiles, but found, that it is slow. Could be, because I created 10.000 and more geometries.

There are only few geometries: point, line, polygon. Are there any other?

**Comment by radderz** [comment]
- It would be nice to have a world wrapping system for the map, i.e. so that we can scroll left for instance infinitely and have the map wrap back on itself.

I have implemented this myself in a pretty hacky way, but this is the way that all other flattened world map systems work like google maps, here maps, osm etc.

I like the idea of supporting NTS, it is a great library and would enable the stripping of a fair bit of code from this project.

I use NTS a lot and have found it a good library.

**Comment by pauldendulk** [comment]
@charlenni That is worrying. I was thinking of making the core less dependent on geometries. This becomes possible by using the rendering callback for everything. This way we do not have to know the type of the geometry in the Renderer class. Implementations of the callback could use whatever they want. For locations and boundingbox I would like to use our own custom types. 

There will still be operations on geometries like projections and such, but they could be inside layers and providers.

Not sure how far we can get with this approach, have to try this in code. So perhaps a geometry type could be a vectortile and the core of Mapsui should never have to know that it is.

**Comment by pauldendulk** [comment]
@radderz ah, a logical suggestion. Something that we Europeans do't see as much a problem. Still I don't think it should have the highest priority right now. I think it is a nasty feature to implement which could have a lot of edge cases that pop up later on. For instance, how would it work with the limiter. What kind of things did you run into?

Vote here for this feature: https://github.com/Mapsui/Mapsui/issues/518

**Comment by pauldendulk** [comment]
@charlenni btw, about NTS not supporting many geometry type. What kind of types do you expect? I think NTS follows the simple features which is a very particular set of geometries, it has, for instance, no circle. But I am not even sure if NTS supports all the other types specified here: file:///C:/Users/paul/Downloads/06-103r4_Implementation_Specification_for_Geographic_Information_-_Simple_feature_access_-_Part_1_Common_Architecture_v1.2.1.pdf

**Comment by charlenni** [comment]
No, I don't say, that NTS isn't supporting many types. I say, that Mapsui doesn't need many types. NTS is a very good library to do all you can imagine, but is this, what Mapsui needs? One of your goals is, to make Mapsui faster. I don't think, that we can reach this goal, if we use such a heavy library like NTS. So the question is, what could NTS provide for Mapsui, that Mapsui needs?

When I see, how long 2.0 needed, than I would say, that we should make a roadmap for 2.1, 2.5 and 3.0.

**Comment by radderz** [comment]
> @radderz ah, a logical suggestion. Something that we Europeans do't see as much a problem. Still I don't think it should have the highest priority right now. I think it is a nasty feature to implement which could have a lot of edge cases that pop up later on. For instance, how would it work with the limiter. What kind of things did you run into?
> 
> Vote here for this feature: #518

I basically hacked the globalmercator projection to be 7 times as wide and as it escapes the >1 world past the center I wrapped back in -1 world without any form of animation, you can't see it. Performance is fine, but it makes it a little trickier when trying to do navigations using animations and you have to have duplicate features if you allow the user to zoom far enough to see >1 full world width. overall works reasonably well even with a large number of features though.

I have customers with assets in multiple continents so it's a bit weird if I don't enable some form of wrapping. I looked into doing it a bit cleaner, with having the tile fetcher aware of it and the feature drawing side of things also aware of it so I don't need duplicates with a hack but never got around to completing it.

**Comment by radderz** [comment]
> No, I don't say, that NTS isn't supporting many types. I say, that Mapsui doesn't need many types. NTS is a very good library to do all you can imagine, but is this, what Mapsui needs? One of your goals is, to make Mapsui faster. I don't think, that we can reach this goal, if we use such a heavy library like NTS. So the question is, what could NTS provide for Mapsui, that Mapsui needs?
> 
> When I see, how long 2.0 needed, than I would say, that we should make a roadmap for 2.1, 2.5 and 3.0.

I have >4000 SVG icons with touch in a small space for things around airports on my app and am using NTS for the under the hood stuff, so would cause me less pain to have mapsui to use it so I don't have to do so many conversions. I use NTS on server side for some pretty heavy calculations like map matching and it's performance is very good.

One of the best parts of NTS is it enables projections and loads of spatial abilities that you'd get out of the box, i.e. buffering a point or line or polygon, intersections, unions etc.

**Comment by pauldendulk** [comment]
> > For locations and boundingbox I would like to use our own custom types.
> 
> Than we have for bounding boxes Extent (BruTiles), Envelope (NTS) and BoundingBox (Mapsui) all for the same thing. That would be very strange. And geometries in NTS use Coordinates whereas Mapsui use Points, which again exist in NTS too.

Some clarification of what I tried to say in [this comment](https://github.com/Mapsui/Mapsui/issues/906#issuecomment-624453893). What I have in mind is that the core of Mapsui is free of proper geometries, we will need a few types. I just had a look at the code and it seems it would suffice to have a Point, BoundingBox and Quad. Within a Layer or Provider you could choose to use a specific geometry implementation. A VectorLayer could use NTS. A TileLayer could use BruTile in the implementation. A VectorTileLayer could use something else. This would not limit the performance that could be achieved by a specific layer. But I need to investigate this further before I ramble on. Not sure how this will work out with MapInfo and the render callback.

You mention the BruTile Extent. In the current implementation this is only used when we directly work with BruTile classes. It is isolated in the TileLayer implementation and not in any other interfaces. It is not even used in the TileLayer interface. I do not see this as a problem. Typically Mapsui could have many different kind of data sources that makes use of a specific library for that data source. If NTS could be used in such a role I don't see that as a problem.

**Comment by pauldendulk** [comment]
> No, I don't say, that NTS isn't supporting many types. I say, that Mapsui doesn't need many types. NTS is a very good library to do all you can imagine, but is this, what Mapsui needs? One of your goals is, to make Mapsui faster. I don't think, that we can reach this goal, if we use such a heavy library like NTS. So the question is, what could NTS provide for Mapsui, that Mapsui needs?

ah, thanks for the clarification.

> When I see, how long 2.0 needed, than I would say, that we should make a roadmap for 2.1, 2.5 and 3.0.

You have a point. It should be a sprint towards the next release. I know we had this conversation before we started on 2.0 as well, and yet, somehow one and a half year has passed 😄

**Comment by charlenni** [comment]
> Some clarification of what I tried to say in [this comment](https://github.com/Mapsui/Mapsui/issues/906#issuecomment-624453893). What I have in mind is that the core of Mapsui is free of proper geometries, we will need a few types. I just had a look at the code and it seems it would suffice to have a Point, BoundingBox and Quad. Within a Layer or Provider you could choose to use a specific geometry implementation. A VectorLayer could use NTS. A TileLayer could use BruTile in the implementation. A VectorTileLayer could use something else. This would not limit the performance that could be achieved by a specific layer. But I need to investigate this further before I ramble on. Not sure how this will work out with MapInfo and the render callback.

You should have two type of points: a screen version and a location version like Point and Coordinate or Point and GeoPoint. Than it is clear, which to use for what.

Perhaps the conversion to NTS would be a first step and leads to a version 2.1? Or is this a to big step?

**Comment by TheProme** [comment]
Those features are awesome. Just wanted to know how long would it take to make offline VectorTiles a possible feature?
I've seen charlenni's library but I doubt I can simply plug it into my project. Or maybe I could, but I surely need guidance for it :(

**Comment by hajmajeboss** [comment]
This looks really promising - I'm definitely for switch to the NTS, the library is awesome, also caching SKPaths could really help the performance side of things. 

I would also suggest looking into map movements - smooth transitions when flinging/swiping, ability to lock rotation when scaling and vice-versa. I have already implemented that in our fork for Android version and it really helps delivering the professional feel. 

Another thing would probably be raster tiles quality improvements - which probably means not scaling them, but rendering them in native resolution. This is definitely tricky and might not be possible at all. I don't know, if it's possible to get tiles of various sizes from WMS servers - eg. keeping 256x256 for PixelDensity = 1 and downloading bigger tiles for bigger pixel densities. Our biggest map provider in Czechia is using OSM based layers and they are crystal sharp - they are either supplying them in native resolution, or rendering them from vector sets right on the device (which would be really slow with current Mapsui, or would it, with custom renderer?). 

Anyway, amazing work with v2.0 and already looking forward for new features and improvements.

**Comment by charlenni** [comment]
@hajmajeboss Regarding map movments and your implementation: It would be helpful to see your code for Android. Perhaps we could transfer it to the other platforms too.

Crystal sharp maps are normally vector maps. They could be drawn at any resolution. That isn't possible with raster maps. Even, if you use native resolution. With this, you would have only integer zoom levels.

I tried to implement a starter for vector maps. You could test it with the [VectorTiles project](https://github.com/charlenni/VectorTiles). All you see is drawn inside of Mapsui. Use the buttons left or right to zoom animated in and out. Best results you get on iOS, I would say. It doesn't implement any text symbols up to now. This is very difficult.

**Comment by hajmajeboss** [comment]
@charlenni I downloaded gesture detectors from OsmSharpMap or something like that or used standard Android detectors (swipe, fling) and Android animations for fling. The code is quite messy and private at my company's Github, I guess. 

I think @pauldendulk or even you said in a thread about fling from 2018, that you want to implement it at SkiaSharp level, not on platform GUI level, if I remember it correctly. 

If there is no way to achieve crystal clear maps with standard raster WMS service, then OK. We don't have vector data providers anyway, at least for now. I need to get a bit more into OSM to see, if we could actually get vector data and possibly draw them in Mapsui, then it might get interesting for us - we don't use OSM at the moment, but our local provider ČUZK, who only has tile WMS servers and does not supply vector data - and the quality is really bad on big screens with 3.0 pixel density - you can barely read text labels if you're not zoomed to native zoom levels.

**Comment by charlenni** [comment]
Ok. We now have our own animations and use them for the AnimatedNavigator. Fling/swipe should easy be possible. @radderz said, that he has it already. So we wait.

If you have your own Mapsui version, you could try to increase the quality of bitmap drawing. It costs time, but the quality increases.

**Comment by radderz** [comment]
> Ok. We now have our own animations and use them for the AnimatedNavigator. Fling/swipe should easy be possible. @radderz said, that he has it already. So we wait.

I could look to implement this as part of the next release if it's desired. I remember I was wondering how we should do this at the mapsui level rather than what I had done using the animation library for xamarin.forms, i.e. the animation timings with things like easings etc. 

The additional features I have built around the Map that work nicely are, 



- double tap on map background zoom in animation (i.e. no feature handled the clicks) SinInOut

- fling, SinOut, I found 1 second worth of fling to be good as if it's too long it's just too much. A better implementation may be to use some form of "friction" to make it feel right.

pretty much everything else was integrated into mapsui for things like rotation around a point for pinch to zoom/rotate, rotation snapping when close to the 0 rotation.

**Comment by radderz** [comment]
> I would also suggest looking into map movements - smooth transitions when flinging/swiping, ability to lock rotation when scaling and vice-versa. I have already implemented that in our fork for Android version and it really helps delivering the professional feel.

When you say rotation lock, do you mean just disabling rotation? You can already control that? The other thing is rotation snapping, where you specify a degrees value for UnSnapRotationDegrees and ReSnapRotationDegrees. I would recommend 30 and 5 respectively to get a nice zoom experience.

**Comment by radderz** [comment]
> Ok. We now have our own animations and use them for the AnimatedNavigator. Fling/swipe should easy be possible. @radderz said, that he has it already. So we wait.

I didn't realise we had our own animations now, so I can try replace the xamarin.forms animator and then integrate it into mapsui instead.

**Comment by charlenni** [comment]
@radderz Sure we have ;-) With easing (see [here](https://github.com/Mapsui/Mapsui/blob/master/Mapsui/Utilities/Easing.cs)). We have also a AnimatedNavigator (see [here](https://github.com/Mapsui/Mapsui/blob/master/Mapsui/AnimatedNavigator.cs)), even when we have to set it more often than I wish. @pauldendulk We should look into this (AnimatedNavigator instead of Navigator when creating Map) too. With this AnimatedNavigator all transitions (translation, rotation, zooming) could be animated. Additionally there is a FlyTo animation. With all this stuff it should be possible to create a fling gesture.

**Comment by radderz** [comment]
@charlenni I have done the changes including extending some of the existing features of the AnimatedNavigator as it's missing some center point animated zooms for instance. Just need time to test and do a pull.

**Comment by charlenni** [comment]
@radderz That sounds great.

**Comment by radderz** [comment]
@charlenni @pauldendulk I recommend we merge the INavigator and AnimatedNavigator but we seperate out the animated versions like this

CenterOn(Point p)
CenterOnAnimated(Point p, int durationMillis = 300)

as then we can update the INavigator to have animatable features without breaking anything existing and then it's explicit when the navigator is animating vs instant.

Doing it like this would mean we can make the default navigator do animations where desired and then INavigator interface can have all the new features directly.

If you agree to this idea, I can do a pull request with this change.

**Comment by charlenni** [comment]
Moved the AnimatedNavigator discussion to #919.

**Comment by jinyuttt** [comment]
I would like 3D maps to be added and expanded, or to create a branch

**Comment by pauldendulk** [comment]
@jinyuttt 3D is not high on my list since it is a big feature and there are so many other basic things to add. The request comes back with some frequency though.

**Comment by pauldendulk** [comment]
@kubaszostak So far I had ignored everything related to WinUI. I am going to read up a bit. Maui did catch my attention. Not sure how it related to WinUI.

**Comment by pauldendulk** [comment]
@kubaszostak Did some reading and watched the [build 2020 intro](https://www.youtube.com/watch?v=1ZBMHFoMVAs). So, WinUI will replace MFC, WinForms, WPF and UWP, but these older platforms will still be in use, so some users may still like to use WPF or UWP. Still, we need to deprecated some of our platforms now and then to reduce the maintenance burden. WPF is still used quite a lot and since it is an older platform I don't expect it to disappear soon. UWP is used a lot less and newer, I assume it is easier for users to replace this with UWP. In conclusion I think it makes most sense to replace the current UWP version with WinUI, as you suggested.

![image](https://user-images.githubusercontent.com/963462/91147866-29a15a80-e6b9-11ea-9e38-68746f7b3320.png)

**Comment by radderz** [comment]
@pauldendulk MAUI will replace Xamarin.Forms basically, so we'll want to start to migrate to that once we get closer to the end of NEXT year so this might be more of a 4.0. It isn't due till .NET 6 so we are a ways away.

**Comment by pauldendulk** [comment]
@radderz Yes, I agree, MAUI is not for 3.0. I just mentioned it because I am aware of it, while it was not clear to me how it relates WinUI.

**Comment by kubaszostak** [comment]
I agree that we should wait for release of WinUI (spring 2021) and MAUI (November 2021) and then check how it works and how it could be utilised by Mapsui.

**Comment by YordanYanakiev** [comment]
SkiaSharp now can be used in Blazor, and probably this opens the road for MapsUI to go full Web as well.

**Comment by memsom** [comment]
> * It would be nice to have a world wrapping system for the map, i.e. so that we can scroll left for instance infinitely and have the map wrap back on itself.
> 
> I have implemented this myself in a pretty hacky way, but this is the way that all other flattened world map systems work like google maps, here maps, osm etc.

Do you happen to have any example code for that hack? It is something I am going to need to implement shortly and I would love to look at how someone else managed to make it work.... I saw you other comment ad that makes sense (hacking the Mercator) but the devil is always in the details.

**Comment by pauldendulk** [comment]
4.0 was now release. For an updated roadmap look here: https://mapsui.com/documentation/roadmap.html

**Comment by memsom** [comment]
> 4.0 was now release. For an updated roadmap look here: https://mapsui.com/documentation/roadmap.html

Are you sure Xamarin should go in 5.0? We use Mapsui extensively in WOF and Xamarin Forms projects, not least because it is cross platform and we have our own compatibility API that irons out the slight differences. We are unlikely to port the current Xamarin code to Maui in any near future, and we want both Vector and continuous world map. I don't want to fork 4.0 and port your changes as you make them, but that might be the only option if you drop Xamarin so quickly.

**Comment by inforithmics** [comment]
Xamarin.Froms has an End of life of May 1, 2024 So in less than a year. The reason is mainly because after that date no Xamarin.Forms applications will be accepted in the iOS Store. Because newer Versions of XCode aren't anymore supported.

I think removing the UWP Support of Mapsui for Xamarin.Forms could lessen the Support burden, because then Mapsui could target at least .NET Standard 2.1.  (.Net Standard 2.1 is supported by Xamarin.Forms ios and Android but not UWP).

Mapsui.Android and Mapsui.iOS already Target .NET 6 so I think they don't need to be removed.

https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin

**Comment by memsom** [comment]
So, from our perspective, the app is Android only, but the mapping is used both for WPF using the WPF support and Android using Xamarin. I wrote the mapping api layer to be able to replace the underlying mapping, so it is not impossible to move mapping engines, but it would be a major pain in the posterior as we only recently got the support fairly stable across multiple OS and products. 

The next generation will be in Maui, but this legacy version has a shelf like of another 3-5 years. We don't use any app stores, this app is always currently side loaded in Android. That is for "reasons".

**Comment by memsom** [comment]
@pauldendulk if supporting Xamarin Forms going forward is mainly a resourcing issue, I would be happy to try to make it continue to work... just let me know. As I said, otherwise I think we'd need to fork anyway if anything we need is added and Xamarin Forms is dropped. It would be less of a nightmare for us to just make the Xamarin Forms support work going forward.

**Comment by pauldendulk** [comment]
@memsom Reviewing and discussing bigger MR requests to an older version is also work for us. I am trying to get rid all older stuff because the issues take a lot of time, and in general complicates our build. So, if you would really would like to add newer functionality to an older version you might need to fork it.

There may be better ways to move forward though. You mention a shelf life of 3-5 years but support for Xamarin.Forms will end in less than a year, surely there is going be trouble one way or another. Instead of backporting it might be easier to invest in upgrading to MAUI or .NET Android (also see this question here https://github.com/Mapsui/Mapsui/discussions/2165).

**Comment by memsom** [comment]
@pauldendulk It looks like we might now have a window to move to MAUI, but it depends a lot on time being sold to management. 

But, hedging bets, it would still be good to have a backup plan.
